File: datastructure/heap.cpp
#include <bits/stdc++.h>

using namespace std;

template<typename T>
class Heap {

public:
    Heap() {
        this->tree.clear();
        this->tree.push_back(0);
    };
    void push(T x) {
        this->tree.push_back(x);
        this->bubble_up(this->tree.size() - 1);
    };
    T pop() {
        if (this->is_empty()) {
            throw "Empty Tree";
        }
        T res = this->tree[1];
        this->tree[1] = this->tree[this->tree.size()-1];
        this->tree.pop_back();
        this->bubble_down(1);
        return res;
    };
    bool is_empty() {
        return this->tree.size() <= 1;
    }

    T top() {
        if (this->is_empty()) {
            throw "Empty Tree";
        }
        return this->tree[1];
    }
private:
    vector<T> tree;
    std::size_t parent(std::size_t index) {
        return index / 2;
    }
    std::size_t left_child(std::size_t index) {
        return 2 * index;
    }
    std::size_t right_child(std::size_t index) {
        return 2 * index + 1;
    }
    void bubble_up(std::size_t index){
        while(index > 1) {
            std::size_t par = this->parent(index);
            if(this->tree[index] > tree[par]) {
                swap(this->tree[index], tree[par]);
                index = par;
            }
            else break;
        }
    }
    void bubble_down(std::size_t index) {
        while (true) {
            std::size_t largest = index;
            std::size_t left = left_child(index);
            std::size_t right = right_child(index);
            if (left < this->tree.size() && this->tree[left] > this->tree[largest]) {
                largest = left;
            }
            if (right < this->tree.size() && this->tree[right] > this->tree[largest]) {
                largest = right;
            }
            if (largest == index) break;
            else {
                swap(this->tree[index], this->tree[largest]);
                index = largest;
            }
        }
    };
};

int main(){

    Heap<int> heap;
    heap.push(10);
    heap.push(101);

    heap.push(11);
    heap.push(12);
    heap.push(1);
    heap.push(0);
    heap.push(110);
    heap.push(1100);
    cout << heap.pop() << endl;
    cout << heap.top() << endl;
    return 0;
}

File: datastructure/heap/Heap.java
package com.company;

import java.util.ArrayList;
import java.util.Collections;

public class Heap<T extends Comparable<T>> {

    private ArrayList<T> tree;

    public Heap() {
        this.tree = new ArrayList<>();
        this.tree.add(null);
    }

    public void add(T x) {
        this.tree.add(x);
        this.bubbleUp(this.tree.size() - 1);
    }

    boolean isEmpty() {
        return this.tree.size() <= 1;
    }

    T remove() {
        if (this.isEmpty()) {
            throw new NullPointerException("Empty Tree");
        }
        T res = this.tree.get(1);
        this.tree.set(1, this.tree.get(this.tree.size()-1));
        this.tree.remove(this.tree.size() - 1);
        this.bubbleDown(1);
        return res;
    };

    T top() {
        if (this.isEmpty()) {
            throw new NullPointerException("Empty Tree");
        }
        return this.tree.get(1);
    }

    private int parent(int index) {
        return index / 2;
    }

    private int leftChild(int index) {
        return 2 * index;
    }

    private int rightChild(int index) {
        return 2 * index + 1;
    }

    private void bubbleUp(int index){
        while(index > 1) {
            int par = this.parent(index);
            if(this.tree.get(index).compareTo(this.tree.get(par)) > 0) {
                Collections.swap(this.tree, index, par);
                index = par;
            }
            else break;
        }
    }
    void bubbleDown(int index) {
        while (true) {
            int largest = index;
            int left = leftChild(index);
            int right = rightChild(index);
            if (left < this.tree.size() && this.tree.get(left).compareTo(this.tree.get(largest)) > 0) {
                largest = left;
            }
            if (right < this.tree.size() && this.tree.get(right).compareTo(this.tree.get(largest)) > 0) {
                largest = right;
            }
            if (largest == index) break;
            else {
                Collections.swap(this.tree, index, largest);
                index = largest;
            }
        }
    };
}
